arangod/RocksDBEngine/RocksDBReplicationCommon.h

```diff
@@ -41,7 +42,18 @@ class RocksDBReplicationResult : public Result {
   bool minTickIncluded() const;
   void includeMinTick();
 
+  // forwarded methods
+  bool ok() const { return _result.ok(); }
+  bool fail() const { return _result.fail(); }
+  int errorNumber() const { return _result.errorNumber(); }
+  std::string errorMessage() const { return _result.errorMessage(); }
+
+  // access methods
+  Result const& result() { return _result; }
+  Result&& stealResult() { return std::move(_result); }
```

Comment:
```suggestion
  Result&& result() && { return std::move(_result); }
```
then you use it like `std::move(repl).result()`

Reply:
I like this, but I would change the result type to `Result` to avoid the possibility of dangling references.

Reply:
@mpoeter I'm not sure I see where the dangling reference could occur. All the usages of this will be of the form e.g.
```
arangodb::Result someFunc() {
  ContainedResult res;
  return std::move(res).result();
}
```
and `arangodb::Result` has appropriate move constructors. If you think it could happen still, could you explain a bit more about how?

Reply:
This usage pattern is fine, but there are other situations than can produce dangling references, e.g., calling `result` on a temporary object:
```
return someFunctionThatReturnsAResult().result();
```
Even if it is unlikely that such code will ever be written, the change does not cost anything - the result is a temporary object (i.e., also an rvalue) and as such suitable for return-value-optimization (RVO).

Reply:
Ah, I think I understand now. Thank you for catching this. I'll make the adjustment.

Reply:
are you sure ? I often see the pattern `Val&& method() && { return std::move(val);} `
https://github.com/facebook/folly/blob/master/folly/futures/Future.h#L186

Reply:
Yes, I am. Contrary to its name, `std::move` does not move anything - it is merely a cast to an rvalue reference. Consider this example:
```cpp
struct Foo {
  Foo() : _data(new int(42)) {}
  std::unique_ptr<int> _data;
  std::unique_ptr<int>&& data() && { return std::move(_data); }
  static Foo build() { return Foo(); } 
};

Foo f;
auto&& data = std::move(f).data();
assert(nullptr != data.get());
assert(f._data.get() == data.get());
```
The two assertions hold, proving that `data` is simply a (rvalue) reference to `f._data`. And that's why this can cause problems with temporary objects as the following example shows:
```cpp
std::unique_ptr<int>&& getData() { return Foo::build().data(); }
auto data2 = getData(); // this crashes!
```
At the time `data2` would be created by executing its move constructor, the temporary `Foo` instance together with its `_data` member is already destroyed, causing a dangling (rvalue) reference to be passed into the move constructor.

Granted - this pattern probably works fine in most situations. But when it comes to correctnes words like "probably" and "most" tend to make me a little uneasy! :wink: